test/utils/validation.test.ts

```diff
@@ -0,0 +1,186 @@
+import { ObjectId } from 'mongodb'
+import {
+  Property,
+  validateObjectId,
+  validateProperties,
+} from 'utils/validation'
+
+describe('validation', () => {
+  describe('validateObjectId', () => {
+    test('valid id returns true', () => {
+      const validObjectId = new ObjectId()
+      const objectIdStr = validObjectId.toString()
+
+      expect(validateObjectId(objectIdStr)).toBe(true)
+    })
+
+    test('empty string returns false', () => {
+      expect(validateObjectId('')).toBe(false)
+    })
+
+    test('object id that gets transformed returns false', () => {
+      const invalidObjectId = 'aaaabbbbcccc'
+
+      // ensure the object id does get transformed and that it is not valid
+      expect(new ObjectId(invalidObjectId).toString()).not.toBe(invalidObjectId)
+      expect(validateObjectId(invalidObjectId)).toBe(false)
+    })
+
+    test('invalid object id returns false', () => {
+      const invalidObjectId = '3'
+
+      expect(validateObjectId(invalidObjectId)).toBe(false)
+    })
+  })
+
+  describe('validateProperties', () => {
+    const validationModel: Property[] = [
+      {
+        key: 'name',
+        types: 'string',
+        required: true,
+      },
+      {
+        key: 'id',
+        types: 'string|number',
+        required: true,
+      },
+      {
+        key: 'optional',
+        types: 'string',
+        required: false,
+      },
+    ]
+
+    test('valid object returns successful ValidationResult', () => {
+      const validObject = {
+        name: 'test',
+        id: 3,
+        optional: 'optional',
+      }
+
+      expect(validateProperties(validationModel, validObject)).toEqual({
+        success: true,
+        message: '',
+      })
+    })
+
+    test('object without optional field set returns successful ValidationResult', () => {
+      const validObject = {
+        name: 'test',
+        id: 3,
+      }
+
+      expect(validateProperties(validationModel, validObject)).toEqual({
+        success: true,
+        message: '',
+      })
+    })
+
+    test('property with multiple types returns successful ValidationResult for all supported types', () => {
+      const validObject = {
+        name: 'test',
+        id: '3',
+      }
+      const validObject2 = {
+        name: 'test',
+        id: 3,
+      }
+
+      expect(validateProperties(validationModel, validObject)).toEqual({
+        success: true,
+        message: '',
+      })
+      expect(validateProperties(validationModel, validObject2)).toEqual({
+        success: true,
+        message: '',
+      })
+    })
+
+    test('missing required field returns unsuccessful ValidationResult', () => {
+      const invalidObject = {
+        id: 3,
+      }
+
+      expect(validateProperties(validationModel, invalidObject)).toEqual({
+        success: false,
+        message: 'Missing Required Properties: name\n',
```

Comment:
Not sure if it is and it's your call, if there is an easy/readable way to not hard-code the 'name' property key in this string, I think that'd be valuable.

Reply:
name with the other instances of hard-coded key names

Reply:
This is not trivial to do